ui/src/combobox.tsx

```diff
@@ -59,54 +60,108 @@ export interface Combobox {
-          wave.args[m.name] = result.map(item => item.text)
-          return result
-        })
-      },
-      onChange = (_e: React.FormEvent<Fluent.IComboBox>, option?: Fluent.IComboBoxOption, _index?: U, value?: S) => {
-        if (!option && value) {
-          const opt: Fluent.IComboBoxOption = { key: value, text: value }
-          setOptions((prevOptions = []) => [...prevOptions, opt])
-          selectOpt({...opt, selected: true})
-        }
-        if (option && isMultiValued) {
-          selectOpt(option)
-        }
-        else {
-          const v = option?.text || value || ''
-          wave.args[m.name] = v
-          setText(v)
-        }
-        if (m.trigger) wave.push()
+export const XCombobox = ({ model: m }: { model: Combobox }) => !m.values ? <ComboboxSingleSelect model={m} /> : <ComboboxMultiSelect model={m} />
+
+const ComboboxSingleSelect = ({ model: m }: { model: Omit<Combobox, 'values'> }) => {
+  const
+    [options, setOptions] = useOptions(m.choices),
+    [selected, setSelected] = React.useState<S | null>(m.value ?? null),
+    skipUseEffectForValue = React.useRef(false),
+    onChange = (_e: React.FormEvent<IComboBox>, option?: IComboBoxOption, _index?: U, value?: S) => {
+      if (!option && value) setOptions((prevOptions = []) => [...prevOptions, { key: value, text: value }])
+      
+      const v = option?.text || value || ''
+      setSelected(v)
+      
+      wave.args[m.name] = v
+      if (m.trigger) wave.push()
+
+      // Hacky: Ensure that next dynamically change to value from Wave app will trigger the useEffect
+      m.value = v
+      // The line above will trigger the useEffect for change in m.value and we need to skip it
+      skipUseEffectForValue.current = true
+    }
+  
+  React.useEffect(() => {
+    wave.args[m.name] = (selected && options.map(o => String(o.key)).includes(selected || '')) ? selected : null
+  }, [m.name, m.value, options, selected])
+
+  // Select value when "value" is set from Wave App dynamically
+  React.useEffect(() => {
+    if (!skipUseEffectForValue.current) setSelected(m.value!)
+    skipUseEffectForValue.current = false
+  }, [m.value])
+
+  return (
+    <Fluent.ComboBox
+      data-test={m.name}
+      label={m.label}
+      placeholder={m.placeholder}
+      options={options}
+      required={m.required}
+      selectedKey={selected}
+      allowFreeform
+      disabled={m.disabled}
+      autoComplete="on"
+      errorMessage={m.error}
+      onChange={onChange}
+    />
+  )
+}
+
+const ComboboxMultiSelect = ({ model: m }: { model: Omit<Combobox, 'value'> }) => {
+  const
+    [options, setOptions] = useOptions(m.choices),
+    [selected, setSelected] = React.useState<S[]>(m.values ?? []),
+    selectOpt = (option: IComboBoxOption) => {
+      setSelected(keys => {
+        const result = option.selected ? [...keys, String(option.key)] : (keys as S[]).filter(key => key !== option.key)
```

Comment:
Unnecessary type casting. The type is properly inferred automatically already.
```suggestion
        const result = option.selected ? [...keys, String(option.key)] : keys.filter(key => key !== option.key)
```